src/lib/hyrise.hpp

```diff
@@ -0,0 +1,39 @@
+#pragma once
+
+#include "boost/container/pmr/memory_resource.hpp"
+#include "concurrency/transaction_manager.hpp"
+#include "storage/storage_manager.hpp"
+#include "utils/plugin_manager.hpp"
+#include "utils/singleton.hpp"
+
+namespace opossum {
+
+class Hyrise : public Singleton<Hyrise> {
+ public:
+  // Resets the Hyrise state by deleting the Plugin-, Storage-, and TransactionManager
+  // and creating new ones. This is used especially in tests and can lead to a lot of
+  // issues if there are still running tasks / threads that want to access a resource.
+  // You should be very sure that this is what you want. Have a look at base_test.hpp
+  // to see the correct order of resetting things.
+  static void reset() { get() = Hyrise{}; }
+
+  PluginManager plugin_manager;
+  StorageManager storage_manager;
+  TransactionManager transaction_manager;
+
+ private:
+  Hyrise() {
+    // The default_memory_resource must be initialized before Plugin-, Storage-, and
+    // TransactionManager so that it is the last object to be destructed. This prevents
+    // e.g. the StorageManager destructor to deallocate a table that is already
+    // destructed by the default_memory_resource destructor.
```

Comment:
Not destructed. The destructor has never been called. The memory is gone, meaning when the destructor is called, its free() calls will fail.

Reply:
I'm struggling a bit to find a good and brief explanation. What about:

> The default_memory_resource must be initialized before Hyrise's members in order to be torn down after them. If a member (e.g., the StorageManager) saves data that was allocated via the default_memory_resource, we want the member to deallocate the data first before the memory resource is destroyed.

Reply:
Proposal:

The default_memory_resource must be initialized before Hyrise's members so that it is destructed after them and remains accessible during their deconstruction. For example, when the StorageManager is destructed, it causes its stored tables to be deconstructed, too. As these might call deallocate on the default_memory_resource, it is important that the resource has not been destructed before. As objects are destructed in the reverse order of their construction, explicitly initializing the resource first means that it is destructed last.